crates/oxc_semantic/src/builder.rs

```diff
@@ -1765,6 +1765,23 @@ impl<'a> SemanticBuilder<'a> {
                 self.function_stack.push(self.current_node_id);
                 self.make_all_namespaces_valuelike();
             }
+            AstKind::FunctionBody(_) => {
+                if let Some(AstKind::ArrowFunctionExpression(arrow)) =
+                    self.nodes.parent_kind(self.current_node_id)
+                {
+                    if arrow.expression {
+                        // (x) => x++
+                        //        ^^^ when the arrow function is an expression,
+                        //            the `x++` is a implicit return,
+                        //            so we need to count `x` as a read reference
+                        self.current_reference_flags |= ReferenceFlags::Read;
+                    } else {
+                        // (x) => () => { x++ }
+                        //                ^^^ avoid break nested arrow function is not an expression
+                        self.current_reference_flags -= ReferenceFlags::Read;
```

Comment:
I've got a feeling this could add problems down-stream... this is why I added an `is_in_return` flag

Reply:
I've tried another way, please review it again